File: samples/akka-sample-sharding-scala/killrweather-fog/src/main/scala/sample/killrweather/fog/WeatherStation.scala
/*
 * Copyright (C) 2009-2024 Lightbend Inc. <https://www.lightbend.com>
 */
package sample.killrweather.fog

import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding.Post
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.SystemMaterializer

import scala.concurrent.Future
import scala.util.Failure
import scala.util.Random
import scala.util.Success

/**
  *  How many weather stations there are? Currently:
  *    "well over 10,000 manned and automatic surface weather stations,
  *    1,000 upper-air stations, 7,000 ships, 100 moored and 1,000 drifting buoys,
  *    hundreds of weather radars and 3,000 specially equipped commercial aircraft
  *    measure key parameters of the atmosphere, land and ocean surface every day.
  *    Add to these some 16 meteorological and 50 research satellites to get an idea
  *    of the size of the global network for meteorological, hydrological and other
  *    geophysical observations."
  *  - https://public.wmo.int/en/our-mandate/what-we-do/observations
  */
private[fog] object WeatherStation {

  type WeatherStationId = String

  sealed trait Command
  case object Sample extends Command
  private final case class ProcessSuccess(msg: String) extends Command
  private final case class ProcessFailure(e: Throwable) extends Command

  /** Starts a device and it's task to initiate reading data at a scheduled rate. */
  def apply(wsid: WeatherStationId, settings: FogSettings, httpPort: Int): Behavior[Command] =
    Behaviors.setup(ctx =>
      new WeatherStation(ctx, wsid, settings, httpPort).running
    )
}

/** Starts a device and it's task to initiate reading data at a scheduled rate. */
private class WeatherStation(context: ActorContext[WeatherStation.Command], wsid: WeatherStation.WeatherStationId, settings: FogSettings, httpPort: Int) {
  import WeatherStation._

  private val random = new Random()

  private val http = {
    import akka.actor.typed.scaladsl.adapter._
    Http(context.system.toClassic)
  }
  private val stationUrl = s"http://${settings.host}:${httpPort}/weather/$wsid"

  def running: Behavior[WeatherStation.Command] = {
    context.log.info(s"Started WeatherStation {} of total {} with weather port {}",
      wsid, settings.weatherStations, httpPort)

    Behaviors.setup[WeatherStation.Command] { context =>
      context.log.debug(s"Started {} data sampling.", wsid)

      Behaviors.withTimers { timers =>
        timers.startSingleTimer(Sample, Sample, settings.sampleInterval)

        Behaviors.receiveMessage {
          case Sample =>
            val value = 5 + 30 * random.nextDouble
            val eventTime = System.currentTimeMillis
            context.log.debug("Recording temperature measurement {}", value)
            recordTemperature(eventTime, value)
            Behaviors.same

          case ProcessSuccess(msg) =>
            context.log.debug("Successfully registered data: {}", msg)
            // trigger next sample only after we got a successful response
            timers.startSingleTimer(Sample, Sample, settings.sampleInterval)
            Behaviors.same

          case ProcessFailure(e) =>
            throw new RuntimeException("Failed to register data", e)
        }
      }
    }
  }

  private def recordTemperature(eventTime: Long, temperature: Double): Unit = {
    implicit val ec = context.executionContext
    implicit val materializer = SystemMaterializer(context.system).materializer

    // we could also use a class and a Json formatter like in the server
    // but since this is the only json we send this is a bit more concise
    import spray.json._
    import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
    val json = JsObject(
      "eventTime" -> JsNumber(eventTime),
      "dataType" -> JsString("temperature"),
      "value" -> JsNumber(temperature)
    )

    val futureResponseBody: Future[String] = http.singleRequest(Post(stationUrl, json))
      .flatMap (res =>
        Unmarshal(res).to[String].map(body =>
          if (res.status.isSuccess()) body
          else throw new RuntimeException(s"Failed to register data: $body")
        )
      )
    context.pipeToSelf(futureResponseBody) {
      case Success(s) => ProcessSuccess(s)
      case Failure(e) => ProcessFailure(e)
    }

  }

}

File: samples/akka-sample-sharding-java/killrweather-fog/src/main/java/sample/killrweather/fog/WeatherStation.java
package sample.killrweather.fog;

import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.*;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.ContentTypes;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.StatusCode;
import akka.http.javadsl.model.StatusCodes;
import akka.http.javadsl.unmarshalling.Unmarshaller;
import akka.serialization.jackson.JacksonObjectMapperProvider;
import akka.stream.SystemMaterializer;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Optional;
import java.util.Random;
import java.util.concurrent.CompletionStage;

public class WeatherStation extends AbstractBehavior<WeatherStation.Command> {

  interface Command {}
  enum Sample implements Command {
    INSTANCE
  }
  private static final class ProcessSuccess implements Command {
    final String msg;
    public ProcessSuccess(String msg) {
      this.msg = msg;
    }
  }
  private static final class ProcessFailure implements Command {
    final Throwable e;
    public ProcessFailure(Throwable e) {
      this.e = e;
    }
  }


  public static Behavior<Command> create(String wsid, FogSettings settings, int httpPort) {
    return Behaviors.setup(context ->
        Behaviors.withTimers(timers ->
          new WeatherStation(context, timers, wsid, settings, httpPort)
        )
    );
  }

  private final String wsid;
  private final FogSettings settings;
  private final TimerScheduler<Command> timers;
  private final Random random = new Random();
  private final Http http;
  private final String stationUrl;
  private final ObjectMapper objectMapper;

  private WeatherStation(ActorContext<Command> context, TimerScheduler<Command> timers, String wsid, FogSettings settings, int httpPort) {
    super(context);
    this.wsid = wsid;
    this.settings = settings;
    this.timers = timers;
    this.http = Http.get(Adapter.toClassic(context.getSystem()));
    stationUrl = "http://" + settings.host + ":" + httpPort + "/weather/" + wsid;
    objectMapper = JacksonObjectMapperProvider.get(getContext().getSystem()).getOrCreate("weather-station", Optional.empty());
    timers.startSingleTimer(Sample.INSTANCE, Sample.INSTANCE, settings.sampleInterval);
  }

  @Override
  public Receive<Command> createReceive() {
    return newReceiveBuilder()
      .onMessageEquals(Sample.INSTANCE, this::onSample)
        .onMessage(ProcessSuccess.class, this::onProcessSuccess)
        .onMessage(ProcessFailure.class, this::onProcessFailure)
        .build();
  }

  private Behavior<Command> onProcessSuccess(ProcessSuccess success) {
    getContext().getLog().info("Successfully registered data: {}", success.msg);
    timers.startSingleTimer(Sample.INSTANCE, Sample.INSTANCE, settings.sampleInterval);
    return this;
  }

  private Behavior<Command> onProcessFailure(ProcessFailure failure) {
    throw new RuntimeException("Failed to register data", failure.e);
  }

  private Behavior<Command> onSample() throws Exception {
    double value = 5 + 30 * random.nextDouble();
    long eventTime = System.currentTimeMillis();
    getContext().getLog().debug("Recording temperature measurement {}", value);
    recordTemperature(eventTime, value);
    return this;
  }

  private void recordTemperature(long eventTime, double value) throws Exception {
    Data data = new Data(eventTime, "temperature", value);

    // FIXME no Java API in Akka HTTP to do this using the marshalling infra, see akka-http#2128
    String json = objectMapper.writeValueAsString(data);
    CompletionStage<String> futureResponseBody =
        http.singleRequest(HttpRequest.POST(stationUrl).withEntity(ContentTypes.APPLICATION_JSON, json))
          .thenCompose(response ->
            Unmarshaller.entityToString().unmarshal(response.entity(), SystemMaterializer.get(getContext().getSystem()).materializer())
              .thenApply(body -> {
                if (response.status().isSuccess())
                  return body;
                else throw new RuntimeException("Failed to register data: " + body);
              })
          );

    getContext().pipeToSelf(futureResponseBody, (response, failure) -> {
      if (failure == null) {
        return new ProcessSuccess(response);
      } else {
        return new ProcessFailure(failure);
      }
    });

  }

}
